client/verta/verta/runtime.py

```
@@ -0,0 +1,237 @@
+
+    For all models deployed in Verta with active endpoints, the :meth:`~verta.registry.VertaModelBase.predict`
+    function of the model is wrapped inside an object of this class by default.
+    Logs collected inside the `predict()` function are exported to S3.
+
+    Parameters
+    ----------
+    validate : bool, default False
+        If true, each individual call to :meth:`~verta.runtime.log`  will
+        verify that the value provided is JSON serializable.
+
+    Raises
+    ------
+    RuntimeError
+        If logging context has been added via :meth:`~verta.runtime.log()`
+        outside the scope of the model's :meth:`~verta.registry.VertaModelBase.predict`
+        method, or outside the scope of this context manager.
+
+    Examples
+    --------
+
+    .. code-block:: python
+       :emphasize-lines: 4,8,11
+
+        import json
+        from verta import runtime
+
+        with runtime.context() as ctx:
+            runtime.log('labels', {'model_type': 'scikit_learn'})
+            output = 'this_is_output'
+            runtime.log('output', output)
+            print json.dumps(ctx.logs())
+
+        # After exiting the context manager:
+        final_log_entry = json.dumps(ctx.logs())
+
+        # Output (both) = {"labels": {"model_type": "scikit_learn"}, "output": "this_is_output"}
+
+    """
+    def __init__(self, validate: Optional[bool] = False):
+        self._validate = validate
+        self._logs_dict = dict()
+
+    def __enter__(self):
+        """
+        Ensure an empty logging context to start and set validation flag.
+        """
+        if bool(_get_thread_logs()):  # If not an empty dict.
+            raise RuntimeError(
+                " cannot overwrite prior logging context. Please ensure all calls"
+                " to runtime.log() are made within the predict() method of your"
+                " model, or inside the scope of an instance of verta.runtime.context()."
```

This error message seems strange to me. It's advising the user to put `runtime.log()` in the correct place, but this is more about the location of `with runtime.context():`, isn't it?

I think a more relevant message to provide here would be about not being allowed to create a new logging context when one is already active (e.g. nesting contexts within each other).

---

To put it another way, if we wait until `with runtime.context():` to tell the user that they called `runtime.log()` in the wrong place, it's already too late! There's no public way for them to recover/access that log they made, and there are no public functions for them to fix their thread state. We should raise that error in `runtime.log()` to prevent the problem from happening in the first place.